[8.0][WIP][l10n_es_tgss_base][l10n_es_tgss_partner] Campos de la TGSS para res.partner, y preparado para reimplementar en otros

yajo opened the pull request:
The base module provides abstract models with fields needed to interact with
Spanish TGSS.

The partner module implements those abstract models in `res.partner`.

If anybody wants, for example, to implement those in `hr.employee`, with
`sed s/res.partner/hr.employee` in a copy of `l10n_es_tgss_partner` and a
little more adjustments should be enough.

Related to https://github.com/OCA/l10n-spain/issues/174.

Commit:
Add modules l10n_es_tgss_base and l10n_es_tgss_partner.

The base module provides abstract models with fields needed to interact with
Spanish TGSS.

The partner module implements those abstract models in `res.partner`.

If anybody wants, for example, to implement those in `hr.employee`, with
`sed s/res.partner/hr.employee` in a copy of `l10n_es_tgss_partner` and a
little more adjustments should be enough.

yajo commented:
Oops this is incomplete, please don't review yet.

Commit:
Follow PEP8 for exception names.

Commit:
Add special situations fields.

Those include affected by gender violence, by terrorism, and disability
percentage.

Add tests too, and fix docs.

Commit:
Update translations.

yajo commented:
Now it's complete.

Runbot fails with a bug in module `l10n_es_aeat_mod303`, not related.

pedrobaeza commented:
:+1:, but please rename it to _l10n_es_tgss_ to simplify things.

Waiting for _l10n_es_tgss_partner_ and _l10n_es_tgss_hr_!

yajo commented:
_l10n_es_tgss_partner_ is included in this PR (there are 2 modules).

IMHO _l10n_es_tgss_base_ seems a good name for a base module that does nothing else.

_l10n_es_tgss_hr_ is not coming for now, but with this basis, copying & modifying _l10n_es_tgss_partner_ should not be very hard for anyone needing it. :wink:

pedrobaeza commented:
Oh, I didn't see that module! Checked also. The name is like `l10n_es_aeat`. We don't put it `l10n_es_aeat_base`. It's redundant.

Commit:
Rename l10n_es_tgss_base to l10n_es_tgss.

yajo commented:
Done :blush:

pedrobaeza commented:
:+1:

Commit:
Simplify exceptions, split contribution account.

Now models with contribution_account have a separate affiliation_number
field because a person hired by a company will have is personal
affiliation number but is affiliated under the hirer's contribution
account.

yajo commented:
Finally, I'm developing the multi-account for companies, please wait.

Commit:
Automatic module name.

Commit:
Allow multiple codes per company.

Commit:
Method _message is not static.

Commit:
Restructure base class to allow multiple codes.

Commit:
Allow multiple codes in company partners.

Allow too to select in which parent code is hired the contact.

Commit:
Fix tests.

rafaelbn commented:
Como hemos dicho, aquí mejor no usemos en inglés porque podría limitar las posibilidades de participación. Como mandó Pedro el otro día correo a odoospain... así hay más gente que se entera y particia (supuestamente). GRacias

Commit:
Remove _NS and .pot.

Commit:
New inheritance system, requiring less code for submodules.

yajo commented:
- Rebasado.
- Las empresas pueden tener varios CCC.
- Los contactos pueden tener un NAF.
- Los contactos pueden estar dados de alta en un CCC.
- La herencia de las clases abstractas se ha simplificado.

yajo commented:
Error del runbot: #285.

Commit:
Log expected contribution account code.

The UI should **not** inform it because that could tempt the user to
replace the code just to save the record, when the main purpose of the
control digit is to avoid bad input. The failure could be in any other
part of the code and then what could seem as a "fix" to the user would
probably be a worse problem than the unability to save the record.

However, for debugging purposes, it is useful to know which code was
expected, so the logs now render it just before the exception.

Commit:
Fix tests.

yajo commented:
¿Una ayudita con Travis?

Commit:
Dumber/smarter study levels.

Now they don't have a number to make them more reusable. Previous
numbers were being borrowed from [this 2012's
XSD](http://empresas.fundaciontripartita.org/schemas/2012/FinalizacionGrupo_Bonificada.xsd),
but [in 2016 those have
changed](http://empresas.fundaciontripartita.org/app/xml_formato2016.aspx).
Study levels are also used in other contexts than training, and each
context has its own codes, that will be implemented in submodules.

But for training we will need the numbers again, so I'm preparing this
to be reused for that upcoming module.

Also default study levels are now more specific, so they become more
compatible for any purpose.

yajo commented:
Si intentáis generar una traducción para el módulo base, os toparéis con https://github.com/odoo/odoo/pull/10685.

pedrobaeza commented:
Puesto que esto no se llegó a terminar, cierro y luego siempre se puede volver a retomar.
